### PR TITLE
Implement ReferenceCell::point_is_inside.

### PR DESCRIPTION
**Pull request by bangerth**
This is the generalized equivalent of GeometryInfo::is_inside_unit_cell().

/rebuild

### CODE REVIEW

**Review by peterrum — APPROVED** [review]

**Comment by peterrum** [comment]
FYI @nfehn

File: include/deal.II/grid/reference_cell.h [hunk 1]
```diff
@@ -450,6 +450,22 @@ class ReferenceCell
...
+   * reference cell. This is useful because in practice we may often not be able
+   * to compute the coordinates of a point in reference coordinates exactly, but
+   * only up to numerical roundoff.
+   *
+   * The tolerance parameter may be less than zero, indicating that the point
+   * should be safely inside the cell.
+   */
+  template <int dim>
+  bool
+  point_is_inside(const Point<dim> &p, const double eps = 0) const;
```
Comment (nfehn) [inline include/deal.II/grid/reference_cell.h:467]: Maybe formulate as question `is_point_inside()`.

File: include/deal.II/grid/reference_cell.h [hunk 2]
```diff
@@ -450,6 +450,22 @@ class ReferenceCell
    * @{
    */
 
+  /**
+   * Return true if the given point is inside the reference cell of the present
+   * space dimension. This function accepts an additional parameter (which
+   * defaults to zero) which
```
Comment (nfehn) [inline include/deal.II/grid/reference_cell.h:456]: Maybe such geometry functions with tolerances should not use default parameters.

I think the documentation does not exactly specify what `eps = 0` means. I would understand the docu as follows: A point lying on the boundary together with `eps=0` will result in a return value of `true`. However, the formulation in the first sentence "if the given point is inside the reference cell ..." goes in a different direction.

I would prefer `tolerance` (often used in `GridTools`) over `eps`. Is it really numerical roundoff in the sense of eps?

Comment (bangerth) [reply include/deal.II/grid/reference_cell.h]: I made those changes.

File: include/deal.II/grid/reference_cell.h (hunk 1 above)
Comment (bangerth) [reply include/deal.II/grid/reference_cell.h]: I don't like `is_point_inside` as a name. It sounds grammatically wrong. 

Would `contains_point` be better? As in
```
  cell->contains_point(p)
```

Comment (nfehn) [reply include/deal.II/grid/reference_cell.h]: Maybe just is_inside(), in analogy to std::is_same, std::is_enum, etc.

(point.is_inside(cell) would be easier to read, but does not make sense from a software design perspective. I assume this is the reason why you repeated „point“ in the function name?)

**Comment by peterrum** [comment]
> What do others think about the naming of the function?

I am fine with `contains_point()`. One would read `cell.contains_point(a)` as `cell containts point a`. If one wants to be shorter, one could also write `contains()`. But nut sure if that becomes ambiguous.

**Comment by nfehn** [comment]
I understand "contain" in the sense of information/data, not geometrically.

**Comment by nfehn** [comment]
Assume you want to know whether a cell (of one triangulation) is enclosed by another triangulation `tria`. The example `tria->contains_cell(cell)` in my opinion clearly demonstrates that the "contains" nomenclature is misleading. Put differently, we should find a wording that is intuitive not only for the cell/point example, but also for tria/cell and other examples.

**Comment by Rombur** [comment]
I don't have a strong opinion on the name. If we want something more general, we could use the same naming convention as [boost](https://www.boost.org/doc/libs/1_78_0/libs/geometry/doc/html/geometry/spatial_indexes/queries.html#geometry.spatial_indexes.queries.spatial_queries). Use `within` if a box is totally inside another box and `intersect` if a boxes is totally or partially inside an other one. So for this function, something like `point_within_cell`.

**Comment by nfehn** [comment]
`cell.point_within_cell(point)` might be quite hard to read (then, we would need a free function like `point_within_cell(point,cell)` in my opinion).

What about `cell.encloses(point)` as one more suggestion ;)

**Comment by bangerth** [comment]
On 3/8/22 05:01, Niklas Fehn wrote:
> I understand "contain" in the sense of information/data, not geometrically.

That's not how native speakers would interpret it, though. "contains" is used 
in sets, for example. I think "contains" is pretty clearly geometric. (There's 
also the (shipping) container, which quite literally "contains" a bunch of 
objects in the same way as a dam contains the water behind it.)

**Comment by bangerth** [comment]
`contains_point()` it is. I like that name better than `point_is_inside()` anyway :-)

So updated!

**Review by peterrum — APPROVED** [review]

**Pull request merged**